### ISSUE TITLE
Adaptative key/magic key/alternate repeat key/... behaviour?

### ISSUE THREAD

**Issue by andreyadrian** [kmonad/kmonad#751]
[_reference_](https://getreuer.info/posts/keyboards/alt-layouts/index.html#which-alt-keyboard-layout-should-i-learn)
![image](https://github.com/kmonad/kmonad/assets/64794163/dd214048-7c4a-4160-95dc-af2657acb6bf)

----

Is there a built in function for that?
For an adaptative key that changes its value depending of last input.

I dont know haskell, but something like: 

```
(defalias
  mk (
    switch (get-last-key()){
        case (q):
            #(u e n)
            return;
        case (e):
            #(u)
            return;
    }
  )
)
```

----

My current approach: 
- The position of letter W is where the Magic Key will be.
- Pressing Q or E, puts a layer on top. postq, poste, respectively.
- Value of Magic Key changes to magicq or magice
```
(defsrc

  q w e

)
```

```
(defalias

;;Modifiers:
q (around (layer-next postq) q)
e (around (layer-next poste) e)
;;...

;;Outputs:
mq #( u e n )
me #( u )
;;...

)
```

```
(deflayer normal

@q XX @e

)
```

```
(deflayer postq

_ @mq _

)
```

```
(deflayer poste

_ @me _

)
```

**Comment by slotThe** [comment]
No, we don't currently have any way of knowing what the last inputs were

**Comment by Alounv** [comment]
So, if my undestanding is correct we have two possible ways to implement this :

1️⃣ - We implement a way to register the last input, and we add a `rept` key that repeat the last key. We would also need some way to create custom `rept` keys to choose alternate key instead of repeating the same previous key.

2️⃣ - We implement a way to translate a `single-key-layer` into many layers that only changes one key.
For example, a bit like proposed above :
```
(defsrc
   q w e
)

(deflayer single-key poste
  u @single n
)
```
This syntax would be translated in two layers (toggled next for `q` and `e`) that would only include one key each (in the location of @single) respectively `u` and `e`.

This way
- typing `q` would toggle next a layer replacing `w` by `u`.
- typing `n` would toggel next a layer replaycing `w` by `n`

Not sure if 2️⃣ is better but at least it does not require us to register the last input and allows to create repeat as well as alternate repeat keys.

**Comment by Alounv** [comment]
Finally I just added other layers it wasn't so bad I did aliases like this
```
    ;; repeat and alternate repeat
    q #(q (layer-next q))
    ...

    ;; repeat only
    i #(i (layer-next i))
    ...
```
and used layers like this  (in the first layer `q` is the position of my repeat key and `a` in the position of my alternate repeat key)
```
;; repeat and alternate repeat

(deflayer q
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
  _ _ _ q _ _ a _ _ _
)

...

;; repeat only layer

(deflayer i
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
  _ _ _ i _ _ _ _ _ _ 
)

...
```

I add some 20 more layer but each feat in four lines so it's ok.

**Comment by AndreyAdrian** [comment]
> One could of course create the layers manually. But in my case it probably means adding something like 15 layers for the repeat keys (letters I repeat most often) and 10 for my alternate-repeat key (the ones I use). So +25 layers in my config. This is why I'll be really nice to implement a shortcut for this.

Thats exactly what I did 1 month 🤣

**Comment by AndreyAdrian** [comment]
> Finally I just added 20 layers (It wasn't so bad because each layer takes only 4 lines).
> 
> I did aliases like this:
> 
> ```
> (defalias
>     ;; repeat and alternate repeat
> 
>     q #(q (layer-next q))
>     ...
> 
>     ;; repeat only
> 
>     i #(i (layer-next i))
>     ...
> )
> ```
> 
> And used layers like this:
> 
> ```
> ;; repeat and alternate repeat
> 
> (deflayer q
>   _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
>   _ _ _ q _ _ a _ _ _
> )
> 
> ...
> 
> ;; repeat only layer
> 
> (deflayer i
>   _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
>   _ _ _ i _ _ _ _ _ _ 
> )
> 
> ...
> ```
> 
> (in the first layer `q` is the position of my repeat key and `a` in the position of my alternate repeat key).


My approach is way worse lol

`
						
						;;Modifiers:
							m (around (layer-next magic-is-ent) m)
							l (around (layer-next magic-is-k) l)
							c (around (layer-next magic-is-y) c)
							p (around (layer-next magic-is-y) p)

							s (around (layer-next magic-is-k) s)
							t (around (layer-next magic-is-ion) t)
							r (around (layer-next magic-is-l) r)
							d (around (layer-next magic-is-y) d)
							y (around (layer-next magic-is-ou) y)
							
							w (around (layer-next magic-is-ith) w)
							z (around (layer-next magic-is-ation) z)
							k (around (layer-next magic-is-l) k)
							j (around (layer-next magic-is-ust) j)
							g (around (layer-next magic-is-y) g)
							
							spc (around (layer-next magic-is-the) spc)
							
						  ;;
						  ;;
							
							u (around (layer-next magic-is-o) u)
							o (around (layer-next magic-is-a) o)
							q (around (layer-next magic-is-uen) q)
							
							n (around (layer-next magic-is-ion) n)
							a (around (layer-next magic-is-tion) a)
							i (around (layer-next magic-is-on) i)
							
							b (around (layer-next magic-is-ecause) b)	
							. (around (layer-next magic-is-slash) .)					
						;;...
						

						;;Outputs:
							mk1 #( e n t )
							mk2 #( k )
							mk3 #( y )
							mk4 #( m e n t )
							mk5 #( l )
							mk6 #( p )
							mk20 #( o u )
							mk7 #( t h e )
							
							mk17 #( s )
							mk19 #( i t h )
							mk14 #( a t i o n )
							mk16 #( u s t )
							
							;;
							
							mk8 #( o )
							mk11 #( a )
							mk9 #( u e n )
							mk10 #( i o n )
							
							mk12 #( o n )
							mk18 #( t i o n )
							mk15 #( e c a u s e )
							mk13 #( / )
						`

and then my layers

`					(deflayer magic-is-ent
						_    _    _    _    _    _    _    _    _    _    _    _    _          _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _   @mk1  _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _               _
						_    _    _              _                   _    _    _    _          _    _    _
					)
					
					(deflayer magic-is-k
						_    _    _    _    _    _    _    _    _    _    _    _    _          _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _   @mk2  _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _               _
						_    _    _              _                   _    _    _    _          _    _    _
					)
					
					(deflayer magic-is-y
						_    _    _    _    _    _    _    _    _    _    _    _    _          _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _   @mk3  _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _               _
						_    _    _              _                   _    _    _    _          _    _    _
					)
					
					(deflayer magic-is-ment
						_    _    _    _    _    _    _    _    _    _    _    _    _          _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _   @mk4  _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _               _
						_    _    _              _                   _    _    _    _          _    _    _
					)
					
					(deflayer magic-is-l
						_    _    _    _    _    _    _    _    _    _    _    _    _          _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _   @mk5  _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _               _
						_    _    _              _                   _    _    _    _          _    _    _
					)

					(deflayer magic-is-p
						_    _    _    _    _    _    _    _    _    _    _    _    _          _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _   @mk6  _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _               _
						_    _    _              _                   _    _    _    _          _    _    _
					)
					
					(deflayer magic-is-the
						_    _    _    _    _    _    _    _    _    _    _    _    _          _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _   @mk7  _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _               _
						_    _    _              _                   _    _    _    _          _    _    _
					)
					
					;;
					
					(deflayer magic-is-o
						_    _    _    _    _    _    _    _    _    _    _    _    _          _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _   @mk8  _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _               _
						_    _    _              _                   _    _    _    _          _    _    _
					)
					
					(deflayer magic-is-uen
						_    _    _    _    _    _    _    _    _    _    _    _    _          _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _   @mk9  _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _               _
						_    _    _              _                   _    _    _    _          _    _    _
					)
					
					(deflayer magic-is-ion
						_    _    _    _    _    _    _    _    _    _    _    _    _          _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _   @mk10  _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _               _
						_    _    _              _                   _    _    _    _          _    _    _
					)
					
					(deflayer magic-is-a
						_    _    _    _    _    _    _    _    _    _    _    _    _          _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _   @mk11  _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _               _
						_    _    _              _                   _    _    _    _          _    _    _
					)
					
					(deflayer magic-is-on
						_    _    _    _    _    _    _    _    _    _    _    _    _          _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _   @mk12  _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _               _
						_    _    _              _                   _    _    _    _          _    _    _
					)
					
					(deflayer magic-is-slash
						_    _    _    _    _    _    _    _    _    _    _    _    _          _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _   @mk13  _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _               _
						_    _    _              _                   _    _    _    _          _    _    _
					)
					
					(deflayer magic-is-ation
						_    _    _    _    _    _    _    _    _    _    _    _    _          _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _   @mk14  _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _               _
						_    _    _              _                   _    _    _    _          _    _    _
					)
					
					(deflayer magic-is-ecause
						_    _    _    _    _    _    _    _    _    _    _    _    _          _    _    _
						_    _    _    _    _    _    _    _    _    _   _    _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _   @mk15  _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _               _
						_    _    _              _                   _    _    _    _          _    _    _
					)
					
					(deflayer magic-is-ust
						_    _    _    _    _    _    _    _    _    _    _    _    _          _    _    _
						_    _    _    _    _    _    _    _    _    _   _    _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _   @mk16  _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _               _
						_    _    _              _                   _    _    _    _          _    _    _
					)
					
					(deflayer magic-is-s
						_    _    _    _    _    _    _    _    _    _    _    _    _          _    _    _
						_    _    _    _    _    _    _    _    _    _   _    _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _   @mk17  _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _               _
						_    _    _              _                   _    _    _    _          _    _    _
					)
					
					
					(deflayer magic-is-tion
						_    _    _    _    _    _    _    _    _    _    _    _    _          _    _    _
						_    _    _    _    _    _    _    _    _    _   _     _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _   @mk18 _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _               _
						_    _    _              _                   _    _    _    _          _    _    _
					)
					
					(deflayer magic-is-ith
						_    _    _    _    _    _    _    _    _    _    _    _    _          _    _    _
						_    _    _    _    _    _    _    _    _    _   _     _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _   @mk19 _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _               _
						_    _    _              _                   _    _    _    _          _    _    _
					)
					
					(deflayer magic-is-ou
						_    _    _    _    _    _    _    _    _    _    _    _    _          _    _    _
						_    _    _    _    _    _    _    _    _    _   _     _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _   @mk20 _    _    _     _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _
						_    _    _    _    _    _    _    _    _    _    _    _    _               _
						_    _    _              _                   _    _    _    _          _    _    _
					)`

**Comment by Alounv** [comment]
Thinking of it maybe the quickest thing that would help would be to have shortcuts for multiple `_`.
For example replacing:
```
(deflayer q
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
  _ _ _ q _ _ a _ _ _
)
```
by
```
(deflayer q 90_ q _ _ a _ _ _)
```

**Comment by jtroo** [comment]
Wow the multi-layer configurations are impressive! Might be easier to have something like [switch on key history](https://github.com/jtroo/kanata/blob/main/docs/config.adoc#key-history) though.

**Comment by Alounv** [comment]
Agree, @jtroo I'm secretly hoping someone will be smart enough to make `kanata` work on macOS, I tried but couldn't make it work fully.

**Comment by Alounv** [comment]
Well I finally switched to kanata and my config file has shrunk from 500 lines to 100 lines (mainly because of `key-history`)

**Comment by jokesper** [comment]
Personally I think using #831 would be an acceptable solution:
```clojure
(defsrc :name magic ?)
(deflayer :source magic @mk2)
(deflayer :source magic @mk3)
```
Not the best solution but far easier than implementing key history and also managable (especially since we don't need all those fallthrough buttons).
